### ISSUE TITLE
Napari UI widgets demo plugin

### ISSUE THREAD

**Issue by Aerendel** [AllenCell/napari-aicssegmentation#12]
**Story**
As a UX designer, I want code examples of all available UI widgets so that I can produce a UI design that is compatible and looks good in Napari

**A/C**
I know this is done when...

- I can view an example of all relevant UI widgets in a single Napari plugin window
- We have selected the underlying UI library that will be used

**Details**

- Goal is to showcase the available standard UI widgets in a single plugin window so that Thao can 1) review the look and feel of all of them and 2) build her own UI prototypes using python code instead of Powerpoint. This will allow to design a more precise and accurate UI and will result in less development work. 
- As part of this work, we should identify whether we want to use `QT` or `magicgui` (whichever seems easier to use and/or prettier). If there is no obvious choice right at the beginning, consider building 2 versions of this demo - one for each library.
- See https://allencellscience.slack.com/archives/C01L7L5M29H/p1613777279189500

List of possible widgets to include (non-exhaustive):
- Textbox
- Drop down list
- Slide bar
- Checkbox
- Button
- Progress bar
- Labels
- Fieldset / Form group
- Expander (expand / collapse form group)

**Comment by schoinh** [comment]
Came across some insight on using Qt vs MagicGUI on Zulip:
https://napari.zulipchat.com/#narrow/stream/212875-general/topic/magicgui.20customization.3F

TL;DR summary:

> If you want to do things like subtle positioning adjustment, coloration, size tweaking, you should probably just use Qt. You can probably get a little of this done by directly modify the native widget in magicgui (like widget.native.setStyleSheet() or widget.native.setMaximumWidth(), etc...) but you may well be better of just building your own Qt layout.
> 
> yeah, magicgui is still young, and the focus is on functionality over form for the moment. It is also being developed in a backend agnostic way (such that these widgets could eventually be put in a browser notebook for example), so the intention is not to have complete parity with the Qt styling abilities. You will always have the greatest flexibility if you just use Qt directly.

So I guess it depends on how much styling/customization we want. We could also mix and match Qt and MagicGUI.

**Comment by sofroniewn** [comment]
> Goal is to showcase the available standard UI widgets in a single plugin window

You might find running our [qt_theme_sample](https://github.com/napari/napari/blob/v0.4.5/napari/_qt/widgets/qt_theme_sample.py) helpful

```python
python -m napari._qt.widgets.qt_theme_sample
```

It produces the following output 

<img width="493" alt="Screen Shot 2021-02-24 at 11 11 03 AM" src="https://user-images.githubusercontent.com/6531703/109053069-2bac9c00-7691-11eb-9300-859eeb359c3f.png">

using standard qt widgets.

> So I guess it depends on how much styling/customization we want. We could also mix and match Qt and MagicGUI.

Indeed. The magic of magicGui is the autogenerated transition from function signatures to widgets, but it comes at the cost of customization. We can also think about providing more widget types either through napari or magicgui if there are some you are missing (say you want a combined slider and edit box) and improving some of our defaults.

Happy to talk about these tradeoffs anytime. It's very exciting to see this issue!!

**Comment by schoinh** [comment]
@sofroniewn Wow that is super helpful, thanks! 

We would absolutely love a combined slider and text edit box option. Actually, at a meeting this morning (@Aerendel I did a 5-min show and tell of the plugin progress at our Animated Cell meeting) we were talking about how we need either a combined slider and text edit box or a combined slider and spinbox. The other issue is that the Qt slider by default doesn't seem to support float values. We can code all of these in but would be really nice to not have to do it :)

**Comment by schoinh** [comment]
@Aerendel There are dozens of available widgets in Qt and most of them are not relevant (for example, font picker, date/time picker, etc.) Can the A/C be adjusted to something like "I can view an example of all relevant UI widgets in a single Napari plugin window"?

**Comment by Aerendel** [comment]
@schoinh Yes, just use best judgement for this. If we need to experiment with more widgets in the future as the UX evolves we can do that later.

**Comment by sofroniewn** [comment]
> Actually, at a meeting this morning (@Aerendel I did a 5-min show and tell of the plugin progress at our Animated Cell meeting) we were talking about how we need either a combined slider and text edit box or a combined slider and spinbox. The other issue is that the Qt slider by default doesn't seem to support float values. We can code all of these in but would be really nice to not have to do it :)

Yes @liaprins-czi is starting to think about this now too. We might end up with some more standard widgets that people can reuse. We're currently exploring also adding the widgets in https://github.com/agoose77/qt-jsonschema-form as part of our own usage in https://github.com/napari/napari/pull/2212 cc @goanpeca @ppwadhwa, and might be able to add some more to magicGUI too cc @tlambert03

**Comment by Aerendel** [comment]
@schoinh add demo of a expand/collapse layout/form group

**Comment by schoinh** [comment]
@Aerendel It seems there's no built-in Qt widget/layout for a collapsible section. Can we skip this until we know for sure we're gonna need one of these?

**Comment by schoinh** [comment]
Discussed objectives with @thao-do and @Aerendel and decided the work done to date is sufficient. We have decided to use Qt over MagicGUI. Qt demo plugin is on branch `feature/demo-qt-widgets`.
